src/odemis/acq/align/autofocus.py

```diff
@@ -317,9 +344,138 @@ def _DoAutoFocus(future, detector, emt, focus, dfbkg, good_focus):
             future._autofocus_state = FINISHED
 
 
+def _DoExhaustiveFocus(future, detector, emt, focus, dfbkg, good_focus, range):
+    """
+    Moves the optical focus through the whole given range, measures the focus
+    level on each position and ends up where the best focus level was found. In
+    case a significant deviation was found while going through the range, it
+    stops and limits the search within a smaller range around this position.
+    future (model.ProgressiveFuture): Progressive future provided by the wrapper
+    detector: model.DigitalCamera or model.Detector
+    emt (None or model.Emitter): In case of a SED this is the scanner used
+    focus (model.Actuator): The optical focus
+    dfbkg (model.DataFlow): dataflow of se- or bs- detector
+    good_focus (float): if provided, an already known good focus position to be
+      taken into consideration while autofocusing
+    range (tuple): if provided, the search of the best focus position is limited
+      within this range
+    returns:
+        (float): Focus position (m)
+        (float): Focus level
+    raises:
+            CancelledError if cancelled
+            IOError if procedure failed
+    """
+    logging.debug("Starting Exhaustive Autofocus...")
+
+    try:
+        # use the .depthOfField on detector or emitter as maximum stepsize
+        avail_depths = (detector, emt)
+        if model.hasVA(emt, "dwellTime"):
+            # Hack in case of using the e-beam with a DigitalCamera detector.
+            # All the digital cameras have a depthOfField, which is updated based
+            # on the optical lens properties... but the depthOfField in this
+            # case depends on the e-beam lens.
+            avail_depths = (emt, detector)
+        for c in avail_depths:
+            if model.hasVA(c, "depthOfField"):
+                dof = c.depthOfField.value
+                break
+        else:
+            logging.debug("No depth of field info found")
+            dof = 1e-6  # m, not too bad value
+        logging.debug("Depth of field is %f", dof)
+
+        # Pick measurement method based on the heuristics that SEM detectors
+        # are typically just a point (ie, shape == data depth).
+        # TODO: is this working as expected? Alternatively, we could check
+        # MD_DET_TYPE.
+        if len(detector.shape) > 1:
+            logging.debug("Using Optical method to estimate focus")
+            Measure = MeasureOpticalFocus
+        else:
+            logging.debug("Using SEM method to estimate focus")
+            Measure = MeasureSEMFocus
+
+        rng = focus.axes["z"].range
+        if not range:
+            # if no range is given, at least use the hw range
+            range = rng
```

Comment:
You cannot have 2 variables named range and rng: it's too confusing which one is for what. Rename the two rng_***